labop_convert/behavior_specialization.py

```diff
@@ -51,113 +50,110 @@ def on_begin(self, execution: labop.ProtocolExecution):
         # Save basic information about the execution record
         node = record.node.lookup()
-        params = input_parameter_map([
-            pv for pv in record.call.lookup().parameter_values
-            if pv.parameter.lookup().property_value.direction == uml.PARAMETER_IN
-            ])
+        params = input_parameter_map(
+            [
+                pv
+                for pv in record.call.lookup().parameter_values
+                if pv.parameter.lookup().property_value.direction
+                == uml.PARAMETER_IN
+            ]
+        )
         params = {p: str(v) for p, v in params.items()}
         node_data = {
             "identity": node.identity,
             "behavior": node.behavior,
-            "parameters" : params
+            "parameters": params,
         }
+        self.update_objects(record)
         self.data.append(node_data)
 
+    def update_objects(self, record: labop.ActivityNodeExecution):
+        """
+        Update the objects processed by the record.
+
+        Parameters
+        ----------
+        record : labop.ActivityNodeExecution
+            A step that modifies objects.
+        """
+        pass
+
     def resolve_container_spec(self, spec, addl_conditions=None):
-        # Attempt to infer container instances using the remote container ontology
-        # server, otherwise use tyto to look it up from a local copy of the ontology
         try:
             from container_api import matching_containers
+
+            if "container_api" not in sys.modules:
+                raise Exception(
+                    "Could not import container_api, is it installed?"
+                )
+
+            if addl_conditions:
+                possible_container_types = matching_containers(
+                    spec, addl_conditions=addl_conditions
+                )
+            else:
+                possible_container_types = matching_containers(spec)
         except:
-            l.warning('Could not import container_api, is it installed?')
-        else:
-            try:
-                if addl_conditions:
-                    possible_container_types = matching_containers(spec, addl_conditions=addl_conditions)
-                else:
-                    possible_container_types = matching_containers(spec)
-                return possible_container_types
-            except Exception as e:
-                l.warning(e)
-
-        # This fallback only works when the spec query is a simple container class/instance formatted in Manchester owl as cont:<container_uri>. Other container constraints / query criteria are not supported
-        l.warning(f'Cannot resolve container specification using remote ontology server. Defaulting to static ontology copy')
-        container_uri = validate_spec_query(spec.queryString)
-        if container_uri.is_instance():
-            possible_container_types = [container_uri]
-        else:
-            possible_container_types = container_uri.get_instances()
```

Comment:
This code which has been reverted will break the labware inference example.  

The code here first attempts to resolve a labware reference using a remote or dev ontology server, but, if it can't find one, will default to using the local, static copy of the container-ontology.ttl.